metadata-ingestion/src/datahub/ingestion/source/thrift.py

```diff
@@ -0,0 +1,141 @@
+        else:
+            raise NotImplementedError()
+
+    def bind_MCPs_from_struct_(
+        self, ctx: ThriftParser.Struct_Context
+    ) -> Generator[MetadataChangeProposalWrapper, None, None]:
+
+        name = ctx.IDENTIFIER().getText()
+        urn = f"urn:li:dataset:(urn:li:dataPlatform:thrift,{name},PROD)"
+        yield MetadataChangeProposalWrapper(
+            entityType="dataset",
+            changeType=ChangeType.UPSERT,
+            aspectName="schemaMetadata",
+            entityUrn=urn,
+            aspect=SchemaMetadata(
+                schemaName=name,
+                platform="urn:li:dataPlatform:thrift",
+                version=0,
+                hash="",
+                platformSchema=OtherSchema(""),
+                fields=[
+                    self.bind_SchemaField_from_field(field) for field in ctx.field()
+                ],
+            ),
+        )
+
+    def bind_SchemaField_from_field(
+        self, ctx: ThriftParser.FieldContext
+    ) -> SchemaField:
+        return SchemaField(
+            fieldPath=ctx.IDENTIFIER().getText(),
+            type=self.bind_SchemaFieldDataType_from_field_type(ctx.field_type()),
+            nativeDataType=self.bind_nativeDataType_from_field_type(ctx.field_type()),
+        )
+
+    def bind_nativeDataType_from_field_type(
+        self, ctx: ThriftParser.Field_typeContext
+    ) -> str:
+        if ctx.base_type():
+            return ctx.getText()
+        else:
+            raise NotImplementedError()
+
+    def bind_SchemaFieldDataType_from_field_type(
+        self, ctx: ThriftParser.Field_typeContext
+    ) -> SchemaFieldDataType:
+        if ctx.base_type():
+            return self.bind_SchemaFieldDataType_from_base_type(ctx.base_type())
+        else:
+            raise NotImplementedError()
+
+    def bind_SchemaFieldDataType_from_base_type(
+        self, ctx: ThriftParser.Base_typeContext
+    ) -> SchemaFieldDataType:
+        return self.bind_SchemaFieldDataType_from_Real_base_type(ctx.real_base_type())
+
+    def bind_SchemaFieldDataType_from_Real_base_type(
+        self, ctx: ThriftParser.Real_base_typeContext
+    ) -> SchemaFieldDataType:
+        if ctx.TYPE_I32() or ctx.TYPE_I64() or ctx.TYPE_DOUBLE():
+            return SchemaFieldDataType(NumberType())
+        elif ctx.TYPE_STRING:
+            return SchemaFieldDataType(StringType())
+        else:
+            raise NotImplementedError()
+
+
+@dataclass
+class ThriftSource(Source):
+    config: ThriftSourceConfig
+    report: SourceReport = field(default_factory=SourceReport)
+
+    @classmethod
+    def create(cls, config_dict, ctx):
+        config = ThriftSourceConfig.parse_obj(config_dict)
+        return cls(ctx, config)
+
+    def parse(
+        self, filename: str
+    ) -> Generator[MetadataChangeProposalWrapper, None, None]:
+        with open(filename) as text_file:
+            # lexer
+
+            lexer = ThriftLexer(InputStream(text_file.read()))
```

Comment:
Are there any exceptions in this block that we want to catch and report?

Reply:
This requirement will be resolved in following diffs.

Reply:
Is it OK to do these here and add found errors to a metric of report: self.report.errors: List[str]?
`    def parse(
        self, filename: str
    ) -> Generator[MetadataChangeProposalWrapper, None, None]:
        try:
            with open(filename) as text_file:
                # lexer

                lexer = ThriftLexer(InputStream(text_file.read()))
                stream = CommonTokenStream(lexer)
                # parser
                parser = ThriftParser(stream)

                tree = parser.document()
                # evaluator
                yield from Binder().bind_MCPs_from_Document(tree)
        except Exception as e:
            self.report.errors.append(f"Error: {e}")`